Workqueue: Add generic versions that are properly typed

alvaroaleman opened the pull request:
This change adds a generic version of the various workqueue types while
retaining compatibility for the existing exported symbols and constructors.
The generic variants are prefixed with `Typed` and the existing ones are
marked as deprecated to nudge people to transition without breaking
them.

The first commit is the changes in the workqueue itself, the second changes the Customresource Discovery Controller to use a typed queue in order to demonstrate it.

/sig api-machinery
/kind cleanup

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

alculquicondor commented:
Was this discussed with SIG API machinery?

While I would like to see this change, it's highly disruptive to ecosystem controllers which use the client libraries, significantly increasing the cost of upgrading them.

alvaroaleman commented:
> While I would like to see this change, it's highly disruptive to ecosystem controllers which use the client libraries, significantly increasing the cost of upgrading them.

Adding the generic type you want for your workqueue in the typedef and constructor is `highly disruptive`? It took me about two hours to do this for all of k/k. Additionally, if you use something like controller-runtime, you won't have to do anything.

aojea commented:
> It took me about two hours to do this for all of k/k

yeah, but there are lot of projects that will be forced to do it, and 2 hours for all the downstream maintainers of k/k that may have conflicts is a considerable change.
Personally,  as a person that maintain a lot of downstream projects and has to "suffer" the revendoring pain I suggest to think if we can do it in a less disruptive way or if the benefit of the change is so worth that we should do it this way

  , I think @soltysh opinion will be interesting too

liggitt commented:
As we found with client-go signature changes in 1.18, rippling out changes that make existing widespread "boilerplate" code no longer compile at all is *very* hard to do across libraries and projects.

I agree the change is ~simple for code you control directly, but ideally people would be able to opt-into the change if they want the benefit of type information, and not be gridlocked if they have library dependencies using workqueues that haven't updated yet.

Is there a way to model this so that existing code doesn't *have* to change to keep compiling and getting a workqueue without type information? For example, defining *new* generic types and redefining the *existing* types in terms of the new generic types?

alvaroaleman commented:
@sttts voiced the same concerns regarding rippling changes in Slack: https://kubernetes.slack.com/archives/C0EG7JC6T/p1712839640970919

So looks like we have to keep the existing signatures in place under the current names and add a new set of signatures under a new set of names. Do ppl have thoughts on what names would be good? Prefix the current ones with `Typed`?

cici37 commented:
/triage accepted

alvaroaleman committed:
Workqueue: Add generic versions that are properly typed

This change adds a generic version of the various workqueue types while
retaining compatibility for the existing exported symbols and constructors.
The generic variants are prefixed with `Typed` and the existing ones are
marked as deprecated to nudge people to transition without breaking
them.

alvaroaleman committed:
Use typed queue in Customresource Discovery Controller

alculquicondor commented:
The last similar thing we did was the introduction of typed pointer functions https://github.com/kubernetes/utils/pull/283

As you can see, we created an alternate package. That could allow us to keep shorter function names. Although I'm not sure what an alternative name could be.

sttts commented on staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go:
```diff
@@ -24,49 +24,63 @@ import (
 // DefaultControllerRateLimiter is a no-arg constructor for a default rate limiter for a workqueue.  It has
 // both overall and per-item rate limiting.  The overall is a token bucket and the per-item is exponential
+//
+// Deprecated: Use DefaultTypedControllerRateLimiter instead.
```
do we also deprecate `RateLimiter`? Why this symbol, but not many others in this change? (am just curious, no strong opinions. My gut feeling is that many of these are super-wide spread in the ecosystem).

alvaroaleman replied:
Just that I missed it. If there are no objections I would like to deprecate all the untyped variants. It will not cause any build failures, but if ppl use linters they will nudge them to start using the typed variants.

sttts commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -47,16 +49,33 @@ type QueueConfig struct {
 }
 
+// NewTyped constructs a new work queue (see the package comment).
+func NewTyped[T comparable]() *TypedType[T] {
```
`TypedType` is stuttering and looks kind of funny. Maybe just `Typed`?

alvaroaleman replied:
Changed it

enj commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -23,60 +23,64 @@ import (
 // Queue is the underlying storage for items. The functions below are always
 // called from the same goroutine.
-type Queue interface {
+type Queue[T comparable] interface {
```
Why does this struct not have the `Typed` approach?  i.e.

```go
type Queue TypedQueue[any]

type TypedQueue[T comparable] interface {
```

alvaroaleman replied:
Because this got merged yesterday only and I went by the assumption that it is okay to break the interface, given that it has no users in k/k and is not in any released version: https://github.com/kubernetes/kubernetes/pull/123347

enj commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -23,60 +23,64 @@ import (
 // DefaultQueue is a slice based FIFO queue.
-func DefaultQueue() Queue {
-	return new(queue)
+func DefaultQueue[T comparable]() Queue[T] {
```
Does this also need the `Typed` approach?

alvaroaleman replied:
Same as the above

sttts commented:
Did another pass. Looks good. From Slack I figure that everybody is on-board with the direction.

/lgtm
/approve
